pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -43,17 +85,61 @@ func init() {
 	aliasFormat = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
 }
 
+func (e *CloudWatchExecutor) AlertQuery(logsClient *cloudwatchlogs.CloudWatchLogs, ctx context.Context, queryContext *tsdb.TsdbQuery) (*cloudwatchlogs.GetQueryResultsOutput, error) {
+	const MaxAttempts = 5
+	const PollPeriod = 500 * time.Millisecond
+
+	queryParams := queryContext.Queries[0].Model
+	startQueryOutput, err := e.executeStartQuery(ctx, logsClient, queryParams, queryContext.TimeRange)
+
+	if err != nil {
+		return nil, err
+	}
+
+	requestParams := simplejson.NewFromAny(map[string]interface{}{
+		"region":  queryParams.Get("region").MustString(""),
+		"queryId": *startQueryOutput.QueryId,
+	})
+
+	ticker := time.NewTicker(PollPeriod)
+	defer ticker.Stop()
+
+	attemptCount := 0
+	for range ticker.C {
+		if res, err := e.executeGetQueryResults(ctx, logsClient, requestParams); err != nil {
+			return nil, err
+		} else if isTerminated(*res.Status) || attemptCount > MaxAttempts {
+			return res, err
+		}
+
+		attemptCount++
+	}
+
+	return nil, nil
+}
+
 func (e *CloudWatchExecutor) Query(ctx context.Context, dsInfo *models.DataSource, queryContext *tsdb.TsdbQuery) (*tsdb.Response, error) {
 	var result *tsdb.Response
 	e.DataSource = dsInfo
-	queryType := queryContext.Queries[0].Model.Get("type").MustString("")
+
+	queryParams := queryContext.Queries[0].Model
+	isLogAlertQuery := queryParams.Get("fromAlert").MustBool(false) && queryParams.Get("mode").MustString("") == "Logs"
+
+	if isLogAlertQuery {
```

Comment:
Why the special casing for alert query? Is it because of the retries? Is that common pattern in other datasources? Cause I would expect the retry logic be somewhere up the chain in alert engine, without need the retry logic per datasource.

Reply:
Yeah this looks weird. Would break alerting support for cloudwatch metrics (I think).

Reply:
It's not because of the retries, but rather because of the nature of Cloudwatch Logs and its querying model.
With Cloudwatch Logs you don't just make a query and receive all the results back as with many other data sources, but rather after making a request you receive an ID and then successively make requests, receiving partial results each time until the query is complete on Cloudwatch's side. For queries from explore and dashboards, these successive requests are each made from the browser so that partial results can be returned for each query, but alerts are executed on the backend, so the logic for handling these long-running queries needs to be implemented there as a special case for alerts.
Hopefully that's a clear answer but let me know and I'll try to explain in a bit more detail.

Reply:
So if I understand the query logic in alerting:
- Alerting engine runs query with 3x retry
- CW start query
- CW tries to get results as this may take longer we try 5 times

Is that right? So the CW retries are not to retry the whole query but just checking if we have the full results yet?

That makes sense but please add a comment with what you wrote here as that may not be apparent right away.

Reply:
:+1: Gotcha. Thanks for the explanation. There should definitely be comments explaining this as it diverges from other data sources.

Reply:
@aocenas Yep, that's right. I'll add a comment.